### PR TITLE
Time - Vicki 

### PR DESCRIPTION
**Pull request by vwhwang**
# Task List
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe in your own words what the Model is doing in Rails | Model handles data and schema. 
Describe in your own words what the Controller is doing in Rails | controls methods and events/requests
Describe in your own words what the View is doing in Rails | translate data into html (browser)
Describe an edge-case controller test you wrote | if invalid id redirect to not found page
What is the purpose of using strong params? (i.e. the params method in the controller) | dry up code and safer
How are Rails migrations related to Rails models? | track all data change 
Describe one area of Rails that are still unclear on | route names and path

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  # get '/tasks', to:'tasks#index', as: 'tasks'
+  # get '/tasks/new', to:'tasks#new', as:'new_task'
+  # post '/tasks', to:'tasks#create'
+  # get '/tasks/:id', to:'tasks#show', as: 'task'
+  # get '/tasks/:id/edit', to:'tasks#edit', as: 'edit_task'
+  # patch '/tasks/:id', to:'tasks#update'
+  # delete '/tasks/:id', to:'tasks#destroy'
+
+
+  patch '/tasks/:id/markcomplete',to:'tasks#markcomplete', as: 'mark_complete'
```
Comment (beccaelenzil) [inline config/routes.rb:17]: Consider how making two separate methods, for instance `mark_complete` and `mark_incomplete` would make this action idempotent.

File: test/controllers/tasks_controller_test.rb [hunk 2]
```diff
@@ -78,47 +76,118 @@
...
+
+      Task.create(name: "water the air plants", description: "or they will die")
+
+      task = Task.first 
+
+      expect {
+        patch mark_complete_path(task.id) 
+      }.must_differ "Task.count" , 0
+
+      expect(Task.first.completed_at).wont_equal nil
```
Comment (beccaelenzil) [inline test/controllers/tasks_controller_test.rb:189]: remember to include an expectation for where it will redirect_to

File: app/views/tasks/show.html.erb [hunk 3]
```diff
@@ -0,0 +1 @@
+<h3> <%= @task.name %> : <%=@task.description%> </h3>
```
Comment (beccaelenzil) [inline app/views/tasks/show.html.erb:2]: Minor note: Consider also showing the completion status of the task in the show view.

File: test/controllers/tasks_controller_test.rb [hunk 4]
```diff
@@ -78,47 +76,118 @@
...
-      # Your code here
+      get edit_task_path(9999)
+
+      must_respond_with :redirect
     end
   end
   
-  # Uncomment and complete these tests for Wave 3
+
   describe "update" do
```
Comment (beccaelenzil) [inline test/controllers/tasks_controller_test.rb:104]: Nice work fleshing out these tests.

**Comment by beccaelenzil** [comment]
# Task List




## Major Learning Goals/Code Review





| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
| At least 6 commits with meaningful commit messages | ✔️
| Routes follow RESTful conventions | ✔️
| Uses named routes (like `_path`) | ✔️
| Creates Models and migrations | ✔️
| Creates styled views | ✔️
| Handles errors like nonexistant tasks | ✔️
| Uses `form_with` to render forms in Rails | ✔️

## Functional Requirements/Manual Testing

| Functional Requirement | yes/no |
| --- | --- |
| Successfully handles index & show | ✔️
| index & show tests pass | ✔️
| Successfully handles: New, Create | ✔️
| New, Create tests pass | ✔️
| Successfully handles: Edit, Update | ✔️
| Successfully handles: Destroy, Task Complete | ✔️

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 5+ in Code Review && 5+ in Functional Requirements | ✔️






## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized | ✅

**Review by beccaelenzil — COMMENTED** [review]
Great work on this project. It's clear that the learning goals for your first rails project were met. You're off to a great start. Keep up the good work.
